### PR TITLE
feat(client): refactor Flash/redux into TS

### PR DESCRIPTION
**Pull request by ShaunSHamilton**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Related to #42256 

This is hopefully a simplification of our current Redux logic, whilst providing the best typing we could hope for.
What this PR might accomplish:
- [x] Refactor `components/Flash/redux` into TypeScript
- [x] Simplify Redux logic for types
- [ ] Simplify Redux logic for actions
- [x] Simplify Redux logic for reducers
- [ ] Aggregate types and initial values for app state across client

### CODE REVIEW

**Comment by moT01** [comment]
I rebased this locally and got a small conflict - so I pushed the changes.

File: client/src/redux/index.js [hunk 1]
```diff
@@ -214,8 +218,7 @@ export const shouldRequestDonationSelector = state => {
 };
 
 export const userByNameSelector = username => state => {
-  const { user } = state[ns];
-  // TODO: Why return a string or empty objet literal?
```
Comment (ShaunSHamilton) [inline client/src/redux/index.js:218]: This comment still applies...

**Review by moT01 — DISMISSED** [review]
I didn't see any problems with this. A PR was merged that changes [some linting config](https://github.com/freeCodeCamp/freeCodeCamp/pull/42818), causing typescript errors in many files, here and elsewhere. I don’t think this PR needs to be held up for that. I looked at the code, didn’t see anything out of place - played with it locally, everything seemed to work. LGTM 👍 🎉

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Just so we're on the same page: the objective here is to

- start replacing `redux-actions`
- type the custom code that replaces it

Is that a fair assessment?

File: client/src/redux/types.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      activeDonations: number;
+    };
+    showDonationModal: boolean;
+    isOnline: boolean;
+    donationFormState: DefaultDonationFormState;
+  };
+}
+
+export interface FlashState {
+  message: ({ id: string } & FlashMessageArg) | Record<string, never>;
```
[the remark this answers is not in the document]
Comment (ShaunSHamilton) [reply client/src/redux/types.ts:29]: Well, this is specific in that:
a) A flash message cannot (should not) be created without `message` and `type` properties.
b) The FlashState can exist as an empty object literal (`Record<string, never>`)

File: client/src/redux/types.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
+import { FlashApp, FlashMessageArg } from '../components/Flash/redux';
+import { MainApp } from '.';
+import { CURRENT_CHALLENGE_KEY } from '../templates/Challenges/redux/current-challenge-saga';
+// const allApps = [FlashApp];
```
[the remark this answers is not in the document]
Comment (ShaunSHamilton) [reply client/src/redux/types.ts:4]: > Is that a fair assessment?

At the end of the day, I would hope to remove more than just `redux-actions`: `observable`, `rxjs`, `redux-saga`

Realistically, I will not get to those in this sprint (TS-focussed migration) - I will try not to make too many unnecessary changes.

_After that, we would just need to get rid of (in order): Redux, Gatsby, and the client_ 😄

**Comment by ShaunSHamilton** [comment]
> Is that a fair assessment?

At the end of the day, I would hope to remove more than just `redux-actions`: `observable`, `rxjs`, `redux-saga`

Realistically, I will not get to those in this sprint (TS-focussed migration) - I will try not to make too many unnecessary changes.

_After that, we would just need to get rid of (in order): Redux, Gatsby, and the client_ 😄

**Review by moT01 — COMMENTED** [review]
I tested this out and looked over the code decently - everything seems to be working. Left one comment. Other than that curiosity, this LGTM 🎉

**Review by moT01 — DISMISSED** [review]
LGTM 🎉

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Just some nitpicks, this looks good :+1:

**Review by ojeytonwilliams — APPROVED** [review]

**Review by moT01 — APPROVED** [review]
LGTM 🎉

**Pull request merged**